Blish HUD/Controls/TextInputBase.cs

```diff
@@ -0,0 +1,718 @@
+                        UserSetCursorIndex(record.Index);
+                        break;
+                    case OperationType.Delete:
+                        if (InsertChars(record.Index, record.Data, out int length)) {
+                            redoStack.MakeInsert(record.Index, length);
+                            UserSetCursorIndex(record.Index + length);
+                        }
+
+                        break;
+                    case OperationType.Replace:
+                        redoStack.MakeReplace(_text, record.Index, record.Length, record.Data.Length);
+                        DeleteChars(record.Index, record.Length);
+                        InsertChars(record.Index, record.Data, out _);
+                        break;
+                }
+            } finally {
+                _suppressRedoStackReset = false;
+            }
+
+            ResetSelection();
+        }
+
+        protected void UserSetCursorIndex(int newIndex) {
+            if (newIndex > _text.Length) {
+                newIndex = _text.Length;
+            }
+
+            if (newIndex < 0) {
+                newIndex = 0;
+            }
+
+            this.CursorIndex = newIndex;
+        }
+
+        protected void ResetSelection() {
+            this.SelectionStart = _selectionEnd = _cursorIndex;
+        }
+
+        protected void UpdateSelection() {
+            this.SelectionEnd = _cursorIndex;
+        }
+
+        protected void UpdateSelectionIfShiftDown() {
+            if (this.IsShiftDown) {
+                UpdateSelection();
+            } else {
+                ResetSelection();
+            }
+        }
+
+        protected abstract void MoveLine(int delta);
+
+        protected void SelectAll() {
+            this.SelectionStart = 0;
+            this.SelectionEnd   = _text.Length;
+        }
+
+        protected float MeasureStringWidth(string text) {
+            if (string.IsNullOrEmpty(text)) return 0;
+
+            var lastGlyph = _font.GetGlyphs(text).Last();
+
+            return lastGlyph.Position.X + (lastGlyph.FontRegion?.Width ?? 0);
+        }
+
+        private string ProcessText(string value) {
+            value = value?.Replace("\r", string.Empty);
+
+            if (!_multiline) {
+                value = value?.Replace("\n", string.Empty);
+            }
+
+            if (value?.Length > _maxLength) {
+                value = value.Substring(0, _maxLength);
+            }
+
+            return value;
+        }
+
+        protected bool SetText(string value, bool byUser) {
+            string prevText = _text;
+
+            value = ProcessText(value);
+
+            if (!SetProperty(ref _text, value)) return false;
+
+            // TODO: Update formatted text?
+
+            if (!byUser) {
+                this.CursorIndex = _selectionStart = _selectionEnd = 0;
+            }
+
+            if (!_suppressRedoStackReset) {
+                _redoStack.Reset();
+            }
+
+            _cursorMoved = true;
```

Comment:
this will be set anyway in the `OnCursorIndexChanged`, or not?

Reply:
Yep!  Removed it.